protocol/daemons/server/types/health_checker.go

```diff
@@ -0,0 +1,218 @@
+
+	return now.Sub(u.mostRecentFailureStreakError.Timestamp())
+}
+
+// SchedulePoll schedules the next poll for the health-checkable service. If the service is stopped, the next poll
+// will not be scheduled. This method is synchronized.
+func (u *healthCheckerMutableState) SchedulePoll(nextPollDelay time.Duration) {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	// Don't schedule a poll if the health checker has been stopped.
+	if u.stopped {
+		return
+	}
+
+	// Schedule the next poll.
+	u.timer.Reset(nextPollDelay)
+}
+
+// Stop stops the health checker. This method is synchronized.
+func (u *healthCheckerMutableState) Stop() {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	// Don't stop the health checker if it has already been stopped.
+	if u.stopped {
+		return
+	}
+
+	u.timer.Stop()
+	u.stopped = true
+}
+
+// healthChecker encapsulates the logic for monitoring the health of a health-checkable service.
+type healthChecker struct {
+	// mutableState is the mutable state of the health checker. Access to these fields is synchronized.
+	mutableState *healthCheckerMutableState
+
+	// healthCheckable is the health-checkable service to be monitored.
+	healthCheckable types.HealthCheckable
+
+	// pollFrequency is the frequency at which the health-checkable service is polled.
+	pollFrequency time.Duration
+
+	// maxAcceptableUnhealthyDuration is the maximum acceptable duration for a health-checkable service to
+	// remain unhealthy. If the service remains unhealthy for this duration, the monitor will execute the
+	// specified callback function.
+	maxAcceptableUnhealthyDuration time.Duration
+
+	// unhealthyCallback is the callback function to be executed if the health-checkable service remains
+	// unhealthy for a period of time greater than or equal to the maximum acceptable unhealthy duration.
+	// This callback function is executed with the error that caused the service to become unhealthy.
+	unhealthyCallback func(error)
+
+	// timeProvider is used to get the current time. It is added as a field so that it can be mocked in tests.
+	timeProvider libtime.TimeProvider
+
+	logger log.Logger
+}
+
+// Poll executes a health check for the health-checkable service. If the service has been unhealthy for longer than the
+// maximum acceptable unhealthy duration, the callback function is executed.
+// This method is publicly exposed for testing. This method is synchronized.
+func (hc *healthChecker) Poll() {
+	err := hc.healthCheckable.HealthCheck()
+	now := hc.timeProvider.Now()
+
+	// Capture healthy response.
+
+	if err == nil {
+		hc.mutableState.ReportSuccess(now)
+	} else {
+		streakDuration := hc.mutableState.ReportFailure(now, err)
+		// If the service has been unhealthy for longer than the maximum acceptable unhealthy duration, execute the
+		// callback function.
+		if streakDuration >= hc.maxAcceptableUnhealthyDuration {
+			hc.unhealthyCallback(err)
+		}
+	}
+
+	// Schedule next poll.
+	hc.mutableState.SchedulePoll(hc.pollFrequency)
```

Comment:
To confirm, we want to schedule the next poll when the daemon is unhealthy right? Makes sense to me, but would be nice to add some comments why this is intended.